### PR TITLE
Add learning page

### PR DESCRIPTION
**Pull request by ThomasKroes**
## Learning center
load_mv
The learning center contains all sorts of learning-related resources; from videos to tutorials and examples. The learning center can be invoked from the help menu or the start page (chalkboard icon):

Invoking the learning center from the help menu:
image

Or from the start page (located in the bottom right):
image

## Sections
<details>
<summary>Videos</summary>

The videos section of the learning page provides a tile-based overview of all sorts of videos related to `ManiVault`. Each tile contains a thumbnail of the video and a short description (see image below).
image

When the user hovers over a video thumbnail, an overlay is shown with a `play` button in the center and three icons in the bottom right (from left to right: `summary`, `date` and `tags`):
image

Icon for viewing the video `summary`:
image

One for viewing the publication `date`:
image

And one for viewing the video `tag(s)`:
image

The database of available videos is located in the [learning center](https://www.manivault.studio/api/learning-center.json) JSON file at the `ManiVault` website. This JSON file is used to populate a `LearningPageVideosModel` with all sorts of video-related information (e.g. video identifier, summary, and tags). The videos can be filtered based on tag(s) and string (e.g. title and summary) using the `LearningPageVideosFilterModel`.

Note: in its current form, the video play button links to the associated video on `YouTube`. Ideally, we would use the `YoutubeVideoDialog`, introduced in this PR. There are however some issues that prevent it from being operational (we are working to solve these).

</details>

<details>
<summary>Tutorials</summary>
The tutorials section is not visible (yet), the reason for this is that there is only one tutorial at the moment on the GitHub IO page; how to install `ManiVault`. It would be odd to add that one I suppose.
</details>

<details>
<summary>Examples</summary>

The examples section is currently not accessible as we do not have an infrastructure yet for hosting the examples etc..
</details>

<details>
<summary>Plugin resources</summary>

### Overview
The plugin resources section contains plugin-related information. It shows loaded plugins with possible additional information such as a readme markdown file or a link to its GitHub repository. The plugins can be filtered based on their type.
image

When the user hovers over a plugin, an overlay is shown when the plugin has additional information.
image

In this case, the plugin has an associated `readme`:
image

When clicked, it shows the `readme` in a modal dialog:
image

If available, it will also show a globe icon which links to the associated repository:
image

### Notes for plugin developers
The plugin repository can be set by overriding the virtual `PluginFactory::getRespositoryUrl()` function. If a valid repository URL is provided, the `read me` URL will automatically be available by calling `PluginFactory::getReadmeMarkdownUrl()` (assuming a `readme.md` is available in the root of the repository). By overriding the virtual `PluginFactory::getReadmeMarkdownUrl()` function, the `read-me` URL can be customized.
</details>

## Miscellaneous
Additionally, three buttons are available in the bottom left that link to the ManiVault `Discord`, `Website`, and `Github Organization`:
image
 
The visibility of the sections is controlled  by toggling the page in the settings popup:
image

## Technical contributions
- [x] Add learning page to `ManiVault`
- [x] Restructured the application source files to support multiple front pages
- [x] Create common page classes: `PageWidget`, `PageWidget`, and `PageContentWidget`
- [x] Use these in `StartPageWidget` and `LearningPageWidget`
- [x] Add `LearningPageVideosFilterModel` class for sorting and filtering the `LearningPageVideosModel`
- [x] Add `LearningPageVideoStyledItemDelegate` for custom video display inside a `QListView`
- [x] Add `LearningPageVideoWidget` class, this is the editor widget for the `LearningPageVideoStyledItemDelegate`
- [x] Add `LearningPageVideosWidget` class  (videos section)
- [x] Add `LearningPageTutorialsWidget` class (tutorials section)
- [x] Add `LearningPageExamplesWidget` class (examples section)
- [x] Add `LearningPagePluginActionWidget` class, which is shown in a hover overlay over the plugin
- [x] Add `LearningPagePluginResourcesWidget` class (plugin resources section)
- [x] Add `LearningPageVideosModel` class (lists videos from the [learning center](https://www.manivault.studio/api/learning-center.json) JSON file)
- [x] Link to the learning center page from the start page
- [x] Link to start page from learning center page
- [x] Adds an `AbstractHelpManager` and a `HelpManager` to the `core` for everything help-related
- [x] Extend the `OptionsAction` with a `tags` view
- [x] Add selection convenience functions to the `OptionsAction`; `selectAll()`, `selectNone()` and `selectInvert()`
- [x] Add `toggleOption(const QString& option)` function to the `OptionsAction`
- [x] Add virtual `getCategoryIcon()` to the `PluginFactory` class, this is used to style the `LearningPagePluginActionWidget`
- [x] Implemented `getIcon()` and `getCategoryIcon()`  in plugin factories where needed
- [x] Added trigger action the the `PluginFactory` for launching the `readme` viewer
- [x] Added trigger action the the `PluginFactory` for launching the `readme` viewer
- [x] Add common ManiVault `StandardItemModel` with commonly used and shared features
- [x] Add common ManiVault `SortFilterProxyModel` with some handy default actions for controlling the sorting and filtering
- [x] Use the above models in some places in the `core`
- [x] Implement `getReadmeMarkdownUrl()` in `core` plugins
- [x] Implement `getRespositoryUrl()` in `core` plugins
- [x] Added placeholder `readme.md` to all `core` plugins
- [x] Added `FileDownloader` utility class (used throughout the learning center)
- [x] Added `getColorAsCssString()` utility function (used throughout the learning center)
- [x] Added `MarkdownDialog` class for viewing `markdown` files (used to view `readme` files in the learning center)
- [x] Added `YouTubeVideoDialog` for viewing learning center videos. At the time of this writing, there are some problems with embedding our own videos, so we currently link to the Youtube site.
## Other small bug fixes
- [x] Do not allow toggling (clicking on the text next to the checkbox) when a `ToggleAction` is disabled
- [x] Narrow down filter criteria in model `match` function in `OptionsAction::isOptionSelected(const QString& option) const`

## Contributions to icon fonts
- [x] It is now possible to show brand icons: `_toDiscordAction.setIcon(Application::getIconFont("FontAwesomeBrands", 6, 4).getIcon("discord"));`
- [x] Each Font Awesome icon font now has three variants; `FontAwesomeBrands`, `FontAwesomeRegular`, and `FontAwesomeSolid` (which is the default)

Note: Ideally I'd like to merge this PR by the 24th of may 2024 at the latest.

### CODE REVIEW

File: ManiVault/src/private/LearningPageContentWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+LearningPageContentWidget::LearningPageContentWidget(QWidget* parent /*= nullptr*/) :
+    PageContentWidget(Qt::Vertical, parent),
+    _videosWidget(this),
+    _tutorialsWidget(this),
+    _examplesWidget(this),
+    _pluginResourcesWidget(this),
+    _showVideosAction(this, "Show videos", true),
+    _showTutorialsAction(this, "Show tutorials (coming soon)", false),
+    _showExamplesAction(this, "Show examples (coming soon)", false),
+    _showPluginResourcesAction(this, "Show plugin resources"),
```
Comment (alxvth) [inline ManiVault/src/private/LearningPageContentWidget.cpp:20]: I think we should toggle this on by default as well, like `_showVideosAction`.

File: ManiVault/src/private/HelpManager.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    //#define HELP_MANAGER_VERBOSE
+#endif
+
+namespace mv
+{
+
+HelpManager::HelpManager() :
+    AbstractHelpManager(),
+    _showLearningCenterAction(this, "Learning center"),
+    _toDiscordAction(this, "Join us..."),
```
Comment (alxvth) [inline ManiVault/src/private/HelpManager.cpp:24]: "Join us..." sounds very ominous... maybe we can use [font-awsome discord icon](https://fontawesome.com/icons/discord) directly?

Comment (ThomasKroes) [reply ManiVault/src/private/HelpManager.cpp:24]: I thought about that too, I had some problems showing the discord icon with FontAwesome. Perhaps we can call it: join us on Discord?

File: ManiVault/src/private/HelpManager.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    _showLearningCenterAction.setIconByName("chalkboard-teacher");
+    _showLearningCenterAction.setToolTip("Go to the learning center");
+
+    _toWebsiteAction.setIconByName("globe");
+    _toWebsiteAction.setDefaultWidgetFlags(TriggerAction::Icon);
+
+    _toWikiAction.setIconByName("book");
+    _toWikiAction.setDefaultWidgetFlags(TriggerAction::Icon);
+
+    _toRepositoryAction.setIconByName("code-branch");
```
Comment (alxvth) [inline ManiVault/src/private/HelpManager.cpp:38]: Similar to above, we could use the [fontawesome github icon](https://fontawesome.com/icons/github) directly here.

File: ManiVault/src/private/LearningPageContentWidget.cpp (hunk 1 above)
Comment (ThomasKroes) [reply ManiVault/src/private/LearningPageContentWidget.cpp:20]: The the tutorials and examples section have no content atm.

**Comment by alxvth** [comment]
When clicking on a link the the Readme of a plugin, the linked webpage opens inside the ManiVault window, replacing the Readme. Is it possible to have those link automatically be opened in a browser?

File: ManiVault/src/private/LearningPageContentWidget.cpp (hunk 1 above)
Comment (alxvth) [reply ManiVault/src/private/LearningPageContentWidget.cpp:20]: I tried to just highlight `_showPluginResourcesAction`:

```
_showPluginResourcesAction(this, "Show plugin resources", true),
```

File: ManiVault/src/private/HelpManager.cpp (hunk 2 above)
Comment (alxvth) [reply ManiVault/src/private/HelpManager.cpp:24]: That's also good

File: ManiVault/src/private/LearningPageContentWidget.cpp (hunk 1 above)
Comment (alxvth) [reply ManiVault/src/private/LearningPageContentWidget.cpp:20]: (The plugin resources didn't show - like in the video - for me and I had to toggle them on.)

**Comment by ThomasKroes** [comment]
> When clicking on a link the the Readme of a plugin, the linked webpage opens inside the ManiVault window, replacing the Readme. Is it possible to have those link automatically be opened in a browser?
Sure, we could add that in addition to having a `readme` modal dialog!

File: ManiVault/src/private/LearningPageContentWidget.cpp (hunk 1 above)
Comment (ThomasKroes) [reply ManiVault/src/private/LearningPageContentWidget.cpp:20]: Ah like so! I get it, yeah sure, I will enable it :)

**Comment by alxvth** [comment]
Works great!

**Review by alxvth — APPROVED** [review]

**Comment by JulianThijssen** [comment]
Will this slow down application start-up in any way, until you click on the learning center?

**Comment by ThomasKroes** [comment]
It have not noticed any slow-downs. I suspect it will most likely not slow down app load time. For instance, the population of the videos database for instance is postponed until the videos section is actually shown:

`void LearningPageVideosWidget::showEvent(QShowEvent* showEvent)
{
    _model.populateFromServer();
}`

**Review by JulianThijssen — APPROVED** [review]

**Comment by ThomasKroes** [comment]
> Alright sounds good.
> 
> On a side note, I think in general we should be a little bit careful with adding extra stuff that impacts the performance, even in small ways, because we are so used to running it on very decent machines. I was trying manivault on my mid-range laptop, and it didn't make a great impression. Would be good if we could focus for a period on improving the current state of the software and make it snappy again.
> 
> So yeah, if it can defer all the loading till people click on it that's perfect.

I agree! I already started profiling to see where we can improve the performance, and there seems to be some low-hanging fruit there. A good candidate feature for 1.1...

**Pull request merged**